services/sso/src/main/java/software.amazon.awssdk.services.sso/internal/SsoProfileCredentialsProviderFactory.java

```diff
@@ -0,0 +1,135 @@
+     * {@link #create(SsoCredentialsProfileProperties, SsoCachedAccessTokenProvider)} is invoked. This wrapper is important
+     * because it ensures the parent credentials provider is closed when the sso credentials provider is no longer needed.
+     */
+    private static final class SsoProfileCredentialsProvider implements AwsCredentialsProvider, SdkAutoCloseable {
+        private final SsoClient ssoClient;
+        private final SsoCredentialsProvider credentialsProvider;
+
+        private SsoProfileCredentialsProvider(SsoCredentialsProfileProperties credentialsProperties,
+                                              SsoCachedAccessTokenProvider provider) {
+            String ssoAccountId = credentialsProperties.ssoAccountId();
+            String ssoRoleName = credentialsProperties.ssoRoleName();
+            String ssoRegion = credentialsProperties.ssoRegion();
+
+            this.ssoClient = SsoClient.builder()
+                                      .credentialsProvider(AnonymousCredentialsProvider.create())
+                                      .region(Region.of(ssoRegion))
+                                      .build();
+
+            GetRoleCredentialsRequest request = GetRoleCredentialsRequest.builder()
+                                                                         .accountId(ssoAccountId)
+                                                                         .roleName(ssoRoleName)
+                                                                         .build();
+
+            GetRoleCredentialsRequestSupplier supplier = new GetRoleCredentialsRequestSupplier(request,
+                                                                                               provider.resolveAccessToken());
+
+            this.credentialsProvider = SsoCredentialsProvider.builder()
+                                                             .ssoClient(ssoClient)
+                                                             .refreshRequest(supplier)
+                                                             .build();
+        }
+
+        @Override
+        public AwsCredentials resolveCredentials() {
+            return this.credentialsProvider.resolveCredentials();
+        }
+
+        @Override
+        public void close() {
+            IoUtils.closeQuietly(credentialsProvider, null);
+            IoUtils.closeQuietly(ssoClient, null);
+        }
+    }
+
+
+    private static final class GetRoleCredentialsRequestSupplier implements Supplier {
+        private final GetRoleCredentialsRequest request;
+        private final String cachedToken;
+
+        GetRoleCredentialsRequestSupplier(GetRoleCredentialsRequest request,
+                                          String cachedToken) {
+            this.request = request;
+            this.cachedToken = cachedToken;
+        }
+
+        @Override
+        public Object get() {
+            return request.toBuilder().accessToken(cachedToken).build();
+        }
```

Comment:
This looks constant. Do we need to even use a supplier here? What if the token expires?

Reply:
This is a debatable part. I used to put the check of whether the token expires here in this supplier, but that led to the lacking of flexibility because customers wouldn't be able to customize their own ways of resolving the access token from the cached token file. However, in my current implmenetation, I took the whole resolving access token part out of the factory(including validating the token file), and found out that this supplier became constant and it doesn't really change with the token. Could you point me out to a better way to handle both of these two here?

Reply:
If you need to be able to reload the token in the credentials provider after it expires, you'll have to query the SsoAccessTokenProvider in this supplier to get the access token.

Reply:
This change is done.